### PR TITLE
Sanitize column names

### PR DESCRIPTION
**Pull request by zahash**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
Is there an open you can link this to?

**Comment by zahash** [comment]
https://github.com/pandas-dev/pandas/issues/31417

**Review by jreback — CHANGES_REQUESTED** [review]
-1 as a method in Dataframe

conceptually not averse as a .str method that is well documented (so follows the convention of working str based Series / Index)

this would also target 1.1

**Comment by zahash** [comment]
so, should I make it df.columns.str.sanitize_column_names() ???

**Comment by jreback** [comment]
not averse to a very welll defined convenience function that is composed of other accessors; .sanitize() might be ok

File: doc/source/whatsnew/v0.25.3.rst [hunk 1]
```diff
@@ -16,6 +16,17 @@ Groupby/resample/rolling
...
+Enhancements
+~~~~~~~~~~~~
+
+Sanitizing column names
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Pandas DataFrame now has a method ``sanitize_column_names()``
+the method changes the column names to lowercase, strips any leading and trailing white
+space, replaces space between words to underscores and only keeps alphanumeric
+(and underscore) characters (:issue:`31417`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.25.3.rst]: this is almost equivalent to "ensures we have a valid python variable name", except for excluding leading-numbers.  is that the (related to) the goal?

Comment (zahash) [reply doc/source/whatsnew/v0.25.3.rst]: I didn't think of it in that way but I guess it applies.

**Comment by zahash** [comment]
where are the source files for ``df.columns.str.*`` ??
so that maybe I can change the method from df.sanitize_column_names() to df.columns.str.sanitize() as "jreback" suggested.

If anyone has any suggestions or changes to the above said plan, feel free to comment

**Pull request closed without merging**